tls/s2n_client_hello.c

```diff
@@ -258,11 +258,26 @@ static int s2n_parse_client_hello(struct s2n_connection *conn)
 }
 
 bool s2n_is_tls_12_self_downgrade_required(struct s2n_connection *conn) {
-    /* RSA PSS is required for TLS 1.3 connections. So if there's a possibility that an RSA Certificate could be picked
-     * by a client connection, then downgrade connection to TLS 1.2 if our libcrypto doesn't support RSA PSS. */
-    return ((conn->mode == S2N_SERVER)
-              && conn->config->is_rsa_cert_configured
-              && !s2n_is_rsa_pss_signing_supported());
+    if ((conn->mode == S2N_SERVER)
+          && (conn->config != NULL)
+          && conn->config->is_rsa_cert_configured
+          && !s2n_is_tls13_fully_supported()) {
+        /* RSA PSS is required for TLS 1.3 connections if RSA is used. So if we are a server that doesn't support RSA PSS,
+         * and there's a possibility that an RSA Certificate could be picked by a client connection, then self-downgrade
+         * connection to TLS 1.2. */
```

Comment:
I haven't closely re-reviewed all the Mutual Auth code, but I don't think this will break any mutual auth use cases (and all the mutual auth unit tests are still passing). 

I think at worst, TLS 1.3 clients and servers that were previously using ECDSA + TLS 1.3 might be downgraded to ECDSA + TLS 1.2, but that downgrade fixes s2n client's compatibility with other TLS servers out in the wild.

Reply:
I more meant that this would currently allow a server to negotiate TLS1.3 if using an ECDSA certificate, because we assume the server chooses the auth method so can avoid RSA. But if the server requires mutual auth, couldn't the client still offer an RSA cert?

Reply:
Nice catch! Yes, it'd be possible for a client to send an RSA PSS certificate to the server if the server requested a client certificate, and have the server not be able to handle it. I added a new check for this case.